src/backend/distributed/executor/executor.c

```diff
@@ -622,16 +629,34 @@ DistributedStatementRequiresRollback(DistributedPlan *distributedPlan)
  * distributed execution. In particular, it releases connections and
  * clears their state.
  */
-void
+static void
 FinishDistributedExecution(DistributedExecution *execution)
 {
-	DistributedPlan *distributedPlan = execution->plan;
+	UnsetCitusNoticeLevel();
+
+	if (execution->operation != CMD_SELECT)
+	{
+		/* prevent copying shards in same transaction */
+		XactModificationLevel = XACT_MODIFICATION_DATA;
```

Comment:
We also do this in CitusExecScan, we should probably remove it from there.

Reply:
We should probably remove it from `CitusExecScan` because of `ExecuteTaskList`?

Reply:
Right